// @sentry/nextjs/build/cjs/common/captureRequestError.js
Object.defineProperty(exports, '__esModule', { value: true });

const core = require('@sentry/core');

/**
 * Reports errors for the Next.js `onRequestError` instrumentation hook.
 *
 * Notice: This function is experimental and not intended for production use. Breaking changes may be done to this funtion in any release.
 *
 * @experimental
 */
function experimental_captureRequestError(
  error,
  request,
  errorContext,
) {
  core.withScope(scope => {
    scope.setSDKProcessingMetadata({
      request: {
        headers: request.headers,
        method: request.method,
      },
    });

    scope.setContext('nextjs', {
      request_path: request.url,
      router_kind: errorContext.routerKind,
      router_path: errorContext.routePath,
      route_type: errorContext.routeType,
    });

    scope.setTransactionName(errorContext.routePath);

    core.captureException(error, {
      mechanism: {
        handled: false,
      },
    });
  });
}

exports.experimental_captureRequestError = experimental_captureRequestError;
