// @liveblocks/react/dist/suspense.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

























































var _chunkBO46HBAIjs = require('./chunk-BO46HBAI.js');

// src/suspense.ts
var _core = require('@liveblocks/core');
var _client = require('@liveblocks/client');
_core.detectDupes.call(void 0, _chunkBO46HBAIjs.PKG_NAME, _chunkBO46HBAIjs.PKG_VERSION, _chunkBO46HBAIjs.PKG_FORMAT);
























































exports.ClientContext = _chunkBO46HBAIjs.ClientContext; exports.ClientSideSuspense = _chunkBO46HBAIjs.ClientSideSuspense; exports.LiveblocksProvider = _chunkBO46HBAIjs.LiveblocksProvider; exports.RoomContext = _chunkBO46HBAIjs.RoomContext; exports.RoomProvider = _chunkBO46HBAIjs._RoomProvider; exports.shallow = _client.shallow; exports.useAddReaction = _chunkBO46HBAIjs._useAddReaction; exports.useBatch = _chunkBO46HBAIjs.useBatch; exports.useBroadcastEvent = _chunkBO46HBAIjs._useBroadcastEvent; exports.useCanRedo = _chunkBO46HBAIjs.useCanRedo; exports.useCanUndo = _chunkBO46HBAIjs.useCanUndo; exports.useClient = _chunkBO46HBAIjs.useClient; exports.useCreateComment = _chunkBO46HBAIjs.useCreateComment; exports.useCreateThread = _chunkBO46HBAIjs._useCreateThread; exports.useDeleteAllInboxNotifications = _chunkBO46HBAIjs.useDeleteAllInboxNotifications; exports.useDeleteComment = _chunkBO46HBAIjs.useDeleteComment; exports.useDeleteInboxNotification = _chunkBO46HBAIjs.useDeleteInboxNotification; exports.useDeleteThread = _chunkBO46HBAIjs._useDeleteThread; exports.useEditComment = _chunkBO46HBAIjs.useEditComment; exports.useEditThreadMetadata = _chunkBO46HBAIjs._useEditThreadMetadata; exports.useErrorListener = _chunkBO46HBAIjs.useErrorListener; exports.useEventListener = _chunkBO46HBAIjs._useEventListener; exports.useHistory = _chunkBO46HBAIjs.useHistory; exports.useInboxNotificationThread = _chunkBO46HBAIjs._useInboxNotificationThread; exports.useInboxNotifications = _chunkBO46HBAIjs.useInboxNotificationsSuspense; exports.useLostConnectionListener = _chunkBO46HBAIjs.useLostConnectionListener; exports.useMarkAllInboxNotificationsAsRead = _chunkBO46HBAIjs.useMarkAllInboxNotificationsAsRead; exports.useMarkInboxNotificationAsRead = _chunkBO46HBAIjs.useMarkInboxNotificationAsRead; exports.useMarkThreadAsRead = _chunkBO46HBAIjs.useMarkThreadAsRead; exports.useMarkThreadAsResolved = _chunkBO46HBAIjs.useMarkThreadAsResolved; exports.useMarkThreadAsUnresolved = _chunkBO46HBAIjs.useMarkThreadAsUnresolved; exports.useMutation = _chunkBO46HBAIjs._useMutation; exports.useMyPresence = _chunkBO46HBAIjs._useMyPresence; exports.useOther = _chunkBO46HBAIjs._useOtherSuspense; exports.useOthers = _chunkBO46HBAIjs._useOthersSuspense; exports.useOthersConnectionIds = _chunkBO46HBAIjs.useOthersConnectionIdsSuspense; exports.useOthersListener = _chunkBO46HBAIjs._useOthersListener; exports.useOthersMapped = _chunkBO46HBAIjs._useOthersMappedSuspense; exports.useRedo = _chunkBO46HBAIjs.useRedo; exports.useRemoveReaction = _chunkBO46HBAIjs.useRemoveReaction; exports.useRoom = _chunkBO46HBAIjs._useRoom; exports.useRoomInfo = _chunkBO46HBAIjs.useRoomInfoSuspense; exports.useRoomNotificationSettings = _chunkBO46HBAIjs.useRoomNotificationSettings; exports.useSelf = _chunkBO46HBAIjs._useSelfSuspense; exports.useStatus = _chunkBO46HBAIjs.useStatus; exports.useStorage = _chunkBO46HBAIjs._useStorageSuspense; exports.useStorageRoot = _chunkBO46HBAIjs._useStorageRoot; exports.useStorageStatus = _chunkBO46HBAIjs.useStorageStatusSuspense; exports.useThreadSubscription = _chunkBO46HBAIjs.useThreadSubscription; exports.useThreads = _chunkBO46HBAIjs._useThreadsSuspense; exports.useUndo = _chunkBO46HBAIjs.useUndo; exports.useUnreadInboxNotificationsCount = _chunkBO46HBAIjs.useUnreadInboxNotificationsCountSuspense; exports.useUpdateMyPresence = _chunkBO46HBAIjs._useUpdateMyPresence; exports.useUpdateRoomNotificationSettings = _chunkBO46HBAIjs.useUpdateRoomNotificationSettings; exports.useUser = _chunkBO46HBAIjs._useUserSuspense;
