// jsm-editor/dist/src/components/editor/Editor.jsx
"use strict";
'use client';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Editor = void 0;
var Theme_1 = __importDefault(require("./plugins/Theme"));
var ToolbarPlugin_1 = __importDefault(require("./plugins/ToolbarPlugin"));
var rich_text_1 = require("@lexical/rich-text");
var LexicalAutoFocusPlugin_1 = require("@lexical/react/LexicalAutoFocusPlugin");
var LexicalComposer_1 = require("@lexical/react/LexicalComposer");
var LexicalRichTextPlugin_1 = require("@lexical/react/LexicalRichTextPlugin");
var LexicalContentEditable_1 = require("@lexical/react/LexicalContentEditable");
var LexicalHistoryPlugin_1 = require("@lexical/react/LexicalHistoryPlugin");
var LexicalErrorBoundary_1 = require("@lexical/react/LexicalErrorBoundary");
var react_1 = __importDefault(require("react"));
// Catch any errors that occur during Lexical updates and log them
// or throw them as needed. If you don't throw them, Lexical will
// try to recover gracefully without losing user data.
function Placeholder() {
    return <div className="editor-placeholder">Enter some rich text...</div>;
}
function Editor() {
    var initialConfig = {
        namespace: 'Editor',
        nodes: [rich_text_1.HeadingNode],
        onError: function (error) {
            console.error(error);
            throw error;
        },
        theme: Theme_1.default,
    };
    return (<LexicalComposer_1.LexicalComposer initialConfig={initialConfig}>
      <div className="editor-container size-full">
        <ToolbarPlugin_1.default />

        <div className="editor-inner h-[1100px]">
          <LexicalRichTextPlugin_1.RichTextPlugin contentEditable={<LexicalContentEditable_1.ContentEditable className="editor-input h-full"/>} placeholder={<Placeholder />} ErrorBoundary={LexicalErrorBoundary_1.LexicalErrorBoundary}/>
          <LexicalHistoryPlugin_1.HistoryPlugin />
          <LexicalAutoFocusPlugin_1.AutoFocusPlugin />
        </div>
      </div>
    </LexicalComposer_1.LexicalComposer>);
}
exports.Editor = Editor;
